### ISSUE TITLE
zlib error using `@jbrowse/react-linear-genome-view` with vite

### ISSUE THREAD

**Issue by garrettjstevens** [GMOD/jbrowse-components#2722]
I made a small test example of `@jbrowse/react-linear-genome-vew` in an app build with [vite](https://vitejs.dev/). Basically I copied over the app code from [here](https://github.com/garrettjstevens/jbrowse-react-linear-genome-view-demo). I had to use the solution described [here](https://medium.com/@ftaioli/using-node-js-builtin-modules-with-vite-6194737c2cd2) to polyfill `stream` (for `@gmod/gff`) and `zlib` (for `@gmod/cram`) (I left out all the other polyfills in that example).

The app builds, but encounters an error trying to display a CRAM track:
```
serverSideRenderedBlock.ts:135 TypeError: Cannot read properties of undefined (reading '-5')
    at Zlib2._binding.onerror (node-modules-polyfills:zlib:357:26)
    at Zlib._error (binding.js:264:8)
    at Zlib._write (binding.js:222:10)
    at Zlib.writeSync (binding.js:171:15)
    at Gunzip.Zlib2._processChunk (node-modules-polyfills:zlib:510:31)
    at zlibBufferSync (node-modules-polyfills:zlib:229:17)
    at Object.gunzipSync (node-modules-polyfills:zlib:173:10)
    at CramFile._uncompress (file.js:247:27)
    at CramFile.readBlock (file.js:298:12)
    at async CramSlice.getBlocks (index.js:208:19)
```
![image](https://user-images.githubusercontent.com/25592344/152869246-18676f13-f96f-4ba1-80dc-f937e67a48d7.png)

This also could be a bug in the `zlib` polyfill, but I'm not really sure.

Code for a repro is here: https://codesandbox.io/s/ylt4t with the app running here: https://ylt4t.sse.codesandbox.io/.

**Comment by cmdcolin** [comment]
Good catch and great work on this

I think if I dig around they are using a custom zlib implementation, might not be as stable as pako

I think it goes to this directory
https://github.com/ionic-team/rollup-plugin-node-polyfills/tree/master/polyfills/zlib-lib

**Comment by Manubi** [comment]
Hi,
thanks for your work and time. @garrettjstevens I ended up doing exactly the same. Should've mentioned that. 
What now? :)

**Comment by garrettjstevens** [comment]
I'm not sure what's the best way forward here. I'm now pretty sure it's a problem with the zlib implementation in the polyfill from [`rollup-plugin-node-polyfills`](https://github.com/ionic-team/rollup-plugin-node-polyfills), but it looks like that project is no longer maintained. I don't know enough about vite to try getting it to use a different polyfill.

**Comment by cmdcolin** [comment]
one other possible approach could be to explicitly use pako in our cram-js module (and possibly other modules that we use it)

there may be a small performance hit for cram-js usage but it might be worth it or could be worked around

**Comment by Manubi** [comment]
Morning,
I am on @cmdcolin side. Imho it should be done on the package side as the package is using it. If I would have to do it with vite, after a while I would have 5 different polyfills only for the used packages. Even if I don't need to support older browsers.

**Comment by cmdcolin** [comment]
can confirm explicit_pako does work

**Comment by cmdcolin** [comment]
a new version of @gmod/cram is published (v1.6.1) that switches to pako for browser bundles, so does not require a zlib polyfill

a similar thing was also added to @gmod/bbi which also used zlib and was published as v1.0.32 so it is possible you can use the code without a zlib polyfill now, but I haven't confirmed that

**Comment by Manubi** [comment]
Just reinstalled the `@jbrowse/react-linear-genome-view` without success. Still doesn't work. 

I do get now the following error as this package is using `react-simple-code-editor`.
<img width="463" alt="CleanShot 2022-02-17 at 14 14 20@2x" src="https://user-images.githubusercontent.com/20935138/154489172-7cb8189f-8473-49cb-940a-546f4b474941.png">

If I add the global this as recommended [here](https://github.com/satya164/react-simple-code-editor/issues/86) the bug is gone but I get another one...
<img width="465" alt="CleanShot 2022-02-17 at 14 37 10@2x" src="https://user-images.githubusercontent.com/20935138/154492834-1a97f665-b426-4b1a-8dff-cb8aeb4a255a.png">

**Comment by cmdcolin** [comment]
I tried using @garrettjstevens example here https://codesandbox.io/s/ylt4t downloaded locally (with the down arrow, removed yarn.lock, re-ran yarn) and it seemed to boot up (still not perfect, noticed the "About track" causes issue, not sure if lazy components are wired up)

do have a reproducible example for the error you see?

**Comment by Manubi** [comment]
Indeed the codesandbox example seems to work now but the example is still using several polyfills which I thought you're trying to avaoid with the update?

**Comment by cmdcolin** [comment]
we certainly want to support alternative build systems such as vite, rollup, etc. but we are not at "polyfill zero" yet. our node modules for example use "Buffer" (not UInt8Array) so at least some node polyfills are needed. our gff parser also uses nodejs stream. there are possibly others like the global polyfill

just to try to demonstrate the codesandbox example a little more this example goes has a working vite "build" step just for reference
https://github.com/cmdcolin/vite-jbrowse-react-linear-genome-view

it uses stream-browserify for the stream polyfill

we can try to work towards less polyfills needed but it is not yet but will take time

**Comment by cmdcolin** [comment]
also note: just weirdly, the "About track" dialog box does not work in dev but does work in production build for that repo

**Comment by cmdcolin** [comment]
This particular issue with the zlib error was fixed with the release of v1.6.1 of @gmod/cram :)

maybe can close for now. feel free to open a new issue if anything comes up

**Comment by cmdcolin** [comment]
we could possibly also improve docs related to vite in our readme's...will consider what can be done
